test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -563,4 +605,224 @@ UCS_TEST_P(test_proto_perf_add_func, intersect_middle)
+            add_func(perf, start, point, UCP_PROTO_PERF_FACTOR_LOCAL_TL,
+                     local_tl_func);
+        }
+        start = point + 1;
+    }
+    return perf;
+}
+
+std::vector<size_t>
+test_proto_perf_aggregate::get_sample_points(perf_ptr_t perf)
+{
+    std::vector<size_t> points;
+
+    auto seg = find_lb(perf, 0);
+    while (seg != NULL) {
+        auto start = ucp_proto_perf_segment_start(seg);
+        auto end   = ucp_proto_perf_segment_end(seg);
+        points.push_back(start);
+        points.push_back(end);
+        points.push_back((start + end) / 2);
+        if (start > 0) {
+            points.push_back(start - 1);
+        }
+        if (end == SIZE_MAX) {
+            break;
+        }
+
+        points.push_back(end + 1);
+        seg = find_lb(perf, end + 1);
+    }
+
+    return points;
+}
+
+std::unique_ptr<test_proto_perf::perf_factors_map_t>
+test_proto_perf_aggregate::expected_aggregate_result(const perf_vec_t &perfs,
+                                                     size_t point)
+{
+    perf_factors_map_t agg;
+    for (auto perf : perfs) {
+        ucp_proto_perf_segment_t *seg = find_lb(perf, point);
+        if ((seg == NULL) || (point < ucp_proto_perf_segment_start(seg))) {
+            return nullptr; // Point not found in one of the perf objects
+        }
+
+        for (int factor_id = 0; factor_id < UCP_PROTO_PERF_FACTOR_LAST;
+             ++factor_id) {
+            auto func = ucp_proto_perf_segment_func(
+                    seg, (ucp_proto_perf_factor_id_t)factor_id);
+            if (!ucs_linear_func_is_zero(func, UCP_PROTO_PERF_EPSILON)) {
+                if (agg.find(factor_id) == agg.end()) {
+                    agg[factor_id] = func;
+                } else {
+                    ucs_linear_func_add_inplace(&agg[factor_id], func);
+                }
+            }
+        }
+    }
+
+    return std::unique_ptr<perf_factors_map_t>(new perf_factors_map_t(agg));
+}
+
+void test_proto_perf_aggregate::test_random_funcs(unsigned num_perfs,
+                                                  unsigned num_segments)
+{
+    std::vector<perf_ptr_t> perfs;
+    std::generate_n(std::back_inserter(perfs), num_perfs,
+                    [num_segments]() -> perf_ptr_t {
+                        return generate_random_perf(num_segments);
+                    });
+
+    std::set<size_t> sample_points;
+    for (auto perf : perfs) {
+        auto perf_points = get_sample_points(perf);
+        std::copy(perf_points.begin(), perf_points.end(),
+                  std::inserter(sample_points, sample_points.begin()));
+    }
+
+    m_perf = aggregate(perfs);
+
+    for (auto point : sample_points) {
+        auto expected_result = expected_aggregate_result(perfs, point);
+        if (expected_result) {
+            expect_perf(point, point, *expected_result.get());
+        } else {
+            expect_empty_range(point, point);
+        }
+    }
+
+    if (testing::Test::HasFailure()) {
```

Comment:
maybe dump start/end list in case of failure to ease analysis?

Reply:
what is the start/end list?

Reply:
each `ucp_proto_perf_segment_t`'s `start`/`end` for each element

Reply:
it's already printed when dumping the perf structure